# Overview

mRNA-seq analysis for Swanda et al., 2023.

# Requirements

All the environment of the analysis steps were devloped as docker images.

Host requirements:

+ Ubuntu 20.04.6 LTS
+ Docker version 25.0.5
+ VScode version 1.80.2 with the following extensions:
  + Dev Containers v0.299.0
  + Docker v1.26.0

# Contents

## Preprocessing

`preprocessing/`

Adapter removal, rRNA mapping, and mRNA mapping of the reads.

All the steps were done in a docker image defined in `preprocessing/dockerfile`.

The FASTQ data together with metadata was prepared as `preprocessing/data`.

A custom reference transcriptome was generated by the code in `preprocessing/ref`.

## DEG analysis

`DEG/`

DEG analysis by edgeR (3.40.1).

This was performed in a docker image defined in `DEG/dockerfile`.

The input count data was generated by a custom python code in `figures/myprep.py`.

## Figures

`figures/`

Generating count data and making figures.

All the steps were done in a remote container of VScode defined in `figures/devcontainer`

The figures were made by a custom python code in `figures/Swanda2023.py`

# Contact

Saori Uematsu: su64@cornell.edu, saorid6e9p6p9@gmail.com

# Reference

Robert V. Swanda, Quanquan Ji, Xincheng Wu, Jingyue Yan, Leiming Dong, Yuanhui Mao, Saori Uematsu, Yizhou Dong, and Shu-Bing Qian. Lysosomal cystine governs ferroptosis sensitivity in cancer via cysteine stress response